Fixed #36192 -- Used semantic HTML for buttons.

Phinart98 opened the pull request:
Refactored the admin interface by replacing buttons written with <a> tags with actual <button> tags to introduce semantic HTML and improve accessibility. Updated styling to match the previous implementation.

#### Trac ticket number

ticket-36192

#### Branch description
Some buttons on the admin page were not accessible because they were using <a> tangs but did not link anywhere. I refactored them to use the correct semantic HTML elements.

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [ ] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [x] I have attached screenshots in both light and dark modes for any UI changes.

#### Screenshots
I changed all the following in the admin interface to use semantic HTML and styled them to look the same.
![add_button_img_after](https://github.com/user-attachments/assets/b756c85a-57db-43bf-b656-7580273140d6)
![close_button_img_after](https://github.com/user-attachments/assets/1823c791-4ef2-4f32-b6dc-8f55609b1ea5)
![cal_button_img_after](https://github.com/user-attachments/assets/f160763f-00d4-4231-bbca-ecc82c66ebff)

Phinart98 commented on django/contrib/admin/static/admin/js/inlines.js:
```diff
@@ -15,18 +15,21 @@
  * See: https://opensource.org/licenses/bsd-license.php
  */
-'use strict';
+"use strict";
```
Yes, I just realised that I applied my editor's formatted on it. I will revert it and make another commit.

sarahboyce commented:
@Phinart98 can you look into the JavaScript test failures? Can you also rebase onto the latest main as we have some recent changes to the selenium test suite we will want to be reflected here :+1:

Phinart98 commented:
> @Phinart98 can you look into the JavaScript test failures? Can you also rebase onto the latest main as we have some recent changes to the selenium test suite we will want to be reflected here 👍

Yes. I am currently looking into the JavaScript tests. Noted. I will also rebase onto main.

Phinart98 commented:
Alright, I will look into it. Thank you😊

sarahboyce commented:
Thank you for the updates!
It looks like the mobile css `django/contrib/admin/static/admin/css/responsive.css` needs some updates
![test_added_stacked_inline_with_collapsed_fields--collapsed--mobile_size](https://github.com/user-attachments/assets/32f0200a-2066-4694-b192-ae72397cd6e3)
I also think the rtl css needs some changes too `django/contrib/admin/static/admin/css/rtl.css`

sarahboyce commented:
You can toggle it! It's how you alert people you want another review :grin:

Phinart98 commented:
Hello @sarahboyce , I have hit a brick wall and would love to hear what you suggest I do. My navigator suggested that I commented on the PR earlier during our final weekly check-in but I was trying different things on my own first😅.

1. 
I made changes to the mobile view and rtl views as suggested but I rebased my branch onto main once again before pushing because I felt like it has been a while since my last commit. After pushing, the selenium tests failed again but this time around, the errors were unusual. They were unreasonably numerous and seemed to have nothing to do with the changes I made. 

My local machine is also not powerful enough to run the selenium tests smoothly so I like to use the CI checks on Github. The Django test suite passes successfully too. To see what could be going wrong, I reverted the change and kept things at the point of your last review to see if the selenium tests will pass. However, they failed again and were not consistent as well. - they were flaky. I tried that a few more times and the same things happened(flaky test failures). Sorry about the many force pushes on the PR currently. I read a lot about rebasing when looking into the problem and realised that it's not a recommended thing to do often because it makes things difficult for reviewers because of the way it messes with commit history.

So now I want to find out from you if there are some insights you could share on what could be going wrong? There is currently no new code to review. I haven't pushed the new changes you requested so that we can focus on the test failures.  Currently, the state of the commits is the point of your last review(where all tests passed).

---

2.
This is separate from no.1 above. It's about the actual changes requested. Can you please share more details about the changes? The only difference I could point out on mobile based on the screenshot you shared is the "Add another XXX" button's padding getting reduced. If there are more changes I'm missing, please give me more details about them.

For RTL views, I couldn't find any issues. I edited the CSS  for the inline labels to stay with their heading text in RTL mode but it wasn't a button issue.  Below are screenshots to demonstrate what I mean:
I changed A below to look like B.
A.
![rtl_wrong](https://github.com/user-attachments/assets/1f5efb1b-65c6-40f9-9d62-a28f4cfab10d)
B.
![rtl_fixed](https://github.com/user-attachments/assets/dc2fed65-f131-43d6-b0f4-dbf147ecbc9c)

Because It has nothing to do with the buttons, should I keep the change or does it deserve a ticket on its own?

anorthall commented:
Hi @Phinart98! Congratulations on making your first PR - sorry to hear about the trouble with rebasing and the Selenium tests. We are here at DjangoCon Europe and @sarahboyce kindly rebased your PR against `main` to fix the issues with the Selenium tests. These were unrelated to your changes and are passing now. Note that you'll need to pull the changes to your local branch before you continue working!

I reviewed your changed and found that you've missed some buttons in the admin which also need updating. I found the following `<a>` tags with the `button` class (which now do not render properly as a result of the changes to CSS). These will need to be updated to `<button>` tags:

- `django/contrib/auth/templates/auth/widgets/read_only_password_hash.html`
- `django/contrib/admin/templates/admin/delete_selected_confirmation.html`
- `django/contrib/admin/templates/admin/delete_confirmation.html`

I imagine you'll also need to update some more tests once you've changed these.

I also found that visually, the buttons are now not the same as they were previously. They are actually now both differently and inconsistently sized. The previous buttons looked like this:

![CleanShot 2025-04-26 at 15 00 33@2x](https://github.com/user-attachments/assets/6a049b50-6788-4ee2-9a07-51f8739aea0d)
![CleanShot 2025-04-26 at 14 56 03@2x](https://github.com/user-attachments/assets/fed729c1-08ac-40bd-aad2-e5aeb88e53d6)

Where as now, after your changes, they look like this:

![CleanShot 2025-04-26 at 14 56 38@2x](https://github.com/user-attachments/assets/4688e723-6884-4c78-a4bd-4e0ba3512d0b)
![CleanShot 2025-04-26 at 15 01 18@2x](https://github.com/user-attachments/assets/1eb50d54-5c09-4f61-8d94-da92bfb5d3da)

You can see that they are inconsistently sized now, and the size is not the same as before.

You can also see the impact of not updating some of the old `<a>` tags in this password change button:

![CleanShot 2025-04-26 at 15 12 53@2x](https://github.com/user-attachments/assets/803e2f25-a80a-4e0c-a3cd-b419cd1e288e)

I hope that gives you enough to think about and work with for now! 😄

Phinart98 commented:
Hello @anorthall , Thank you so much for the review and comprehensive comments. I appreciate you catching all that. It's insightful. I will begin working on them as soon as possible.

Thank you @sarahboyce for helping fix the failing Selenium tests.

tut-tuuut reviewed (commented):
Hello :)

We were talking about removing some `role="button"` on anchors and Sarah redirected to here. :) 

I took a look at the code and the admin after rebasing your branch on main (on my local installation).

I could fix a few styling issues (see my suggestions).

About the latest suggestions of Andrew, I don't agree with everything: in `read_only_password_hash.html` it's still a `<a>` but you should remove role=button. In both other templates he mentioned, these should definitely be buttons.

Let us know if you need some help on these. I can take over, too, if needed.

Antoliny0919 commented on django/contrib/admin/static/admin/js/inlines.js:
```diff
@@ -356,4 +356,4 @@
         });
     });
-}
+}
```
I think a new line is needed here!

Antoliny0919 reviewed (commented):
Thank you for working on this ticket @Phinart98 .

I was also discussing the removal of anchor tags with `role="button"`, and Sarah directed me here.

In my opinion, not only the shortcuts inside the calendar and datetime widgets, but also the other buttons within them should be changed from anchor tags to proper <button> elements.
```diff --git a/django/contrib/admin/static/admin/css/base.css b/django/contrib/admin/static/admin/css/base.css
index 2682817a98..defafa9d13 100644
--- a/django/contrib/admin/static/admin/css/base.css
+++ b/django/contrib/admin/static/admin/css/base.css
@@ -535,7 +535,7 @@ select[multiple] {
 
 /* FORM BUTTONS */
 
-.button, input[type=submit], input[type=button], .submit-row input, button.button {
+.button, input[type=submit], input[type=button], .submit-row input, a.button {
     background: var(--button-bg);
     padding: 10px 15px;
     border: none;
diff --git a/django/contrib/admin/static/admin/css/widgets.css b/django/contrib/admin/static/admin/css/widgets.css
index e295d33f76..46174f0c39 100644
--- a/django/contrib/admin/static/admin/css/widgets.css
+++ b/django/contrib/admin/static/admin/css/widgets.css
@@ -472,7 +472,7 @@ span.clearable-file-input label {
     border-bottom: none;
 }
 
-.calendar td.selected a {
+.calendar td.selected button {
     background: var(--secondary);
     color: var(--button-fg);
 }
@@ -481,28 +481,32 @@ span.clearable-file-input label {
     background: var(--darkened-bg);
 }
 
-.calendar td.today a {
+.calendar td.today button {
     font-weight: 700;
     background: var(--secondary);
     color: var(--button-fg);
 
 }
 
-.calendar td a, .timelist a {
+.calendar td button, .timelist button {
     display: block;
+    width: 100%;
+    height: 100%;
+    background: none;
+    border: none;
     font-weight: 400;
     padding: 6px;
     text-decoration: none;
-.calendar td a:focus, .timelist a:focus,
-.calendar td a:hover, .timelist a:hover {
+.calendar td button:focus, .timelist button:focus,
+.calendar td button:hover, .timelist button:hover {
     background: var(--primary);
     color: white;
 }
 
-.calendar td a:active, .timelist a:active {
+.calendar td button:active, .timelist button:active {
     background: var(--header-bg);
     color: white;
 }
diff --git a/django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js b/django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js
index 9da2221f1e..0e136ebab1 100644
--- a/django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js
+++ b/django/contrib/admin/static/admin/js/admin/DateTimeShortcuts.js
@@ -164,7 +164,7 @@
             // where name is the name attribute of the <input>.
             const name = typeof DateTimeShortcuts.clockHours[inp.name] === 'undefined' ? 'default_' : inp.name;
             DateTimeShortcuts.clockHours[name].forEach(function(element) {
-                const time_link = quickElement('a', quickElement('li', time_list), gettext(element[0]), 'role', 'button', 'href', '#');
+                const time_link = quickElement('button', quickElement('li', time_list), gettext(element[0]), 'type', 'button');
                 time_link.addEventListener('click', function(e) {
                     e.preventDefault();
             const cancel_p = quickElement('p', clock_box);
             cancel_p.className = 'calendar-cancel';
-            const cancel_link = quickElement('a', cancel_p, gettext('Cancel'), 'role', 'button', 'href', '#');
+            const cancel_link = quickElement('button', cancel_p, gettext('Cancel'), 'type', 'button');
             cancel_link.addEventListener('click', function(e) {
                 e.preventDefault();
                 DateTimeShortcuts.dismissClock(num);
diff --git a/django/contrib/admin/static/admin/js/calendar.js b/django/contrib/admin/static/admin/js/calendar.js
index b13242078a..78276312c8 100644
--- a/django/contrib/admin/static/admin/js/calendar.js
+++ b/django/contrib/admin/static/admin/js/calendar.js
@@ -160,7 +160,7 @@ depends on core.js for utility functions like removeChildren or quickElement
                 }
 
                 const cell = quickElement('td', tableRow, '', 'class', todayClass);
-                const link = quickElement('a', cell, currentDay, 'role', 'button', 'href', '#');
+                const link = quickElement('button', cell, currentDay, 'type', 'button');
                 link.addEventListener('click', calendarMonth(year, month));
                 currentDay++;
             }
```

As mentioned above, I also changed the buttons inside the calendar and datetime widgets to <button> elements and did my best to preserve their original styling.

I hope this helps. Cheering you on 😃

tut-tuuut commented on django/contrib/admin/static/admin/js/inlines.js:
```diff
@@ -50,16 +50,16 @@
             }
             addButton.on('click', addInlineClickHandler);
-        };
+        };        
```
Perfect ❤️ 
I'm happy to help, feel free to @ me anytime.

thibaudcolas commented:
Static demo site with the changes in this PR: [Static admin index PR 19283](https://preview-19283--django-admin-tests.netlify.app/django_admin_tests/latest/english/admin/)

Antoliny0919 commented:
Hello @Phinart98 :), if you’re busy, would it be okay if I take over the work?

Antoliny0919 commented:
Superseded by https://github.com/django/django/pull/19995
